### PR TITLE
Memoize & unMemoize and apply

### PR DESCRIPTION
**Pull request by crowelch**
...and semicolons.

### CODE REVIEW

File: lib/index.js [hunk 1]
```diff
@@ -462,6 +464,31 @@ exports.timesSeries = function(times, task) {
 	});
 };
 
+exports.memoize = function(promiseFn, hasher) {
+	var callbackFn = constructIterator(promiseFn, 1, true, false);
```
Comment (crispy1989) [inline lib/index.js]: Does the function passed to memoize() support numbers of arguments other than 1?

File: lib/index.js [hunk 2]
```diff
@@ -462,6 +464,31 @@ exports.timesSeries = function(times, task) {
...
+	var memoizedFn = async.memoize(callbackFn);
```
Comment (crispy1989) [inline lib/index.js]: Why is `hasher` not passed in?

File: lib/index.js [hunk 3]
```diff
@@ -462,6 +464,31 @@ exports.timesSeries = function(times, task) {
...
+
```
Comment (crispy1989) [inline lib/index.js]: From the async docs:

`The cache of results is exposed as the memo property of the function returned by memoize.`

It doesn't look like this is supported.

File: test/tests.js [hunk 4]
```diff
@@ -1053,6 +1070,30 @@ describe('pasync', function() {
...
 	});
 
+	it('memoize', function(done) {
+		var myFunc = function(up, down, no) {
+			return Promise.resolve(up * down * no);
+		};
+		var myMemo = pasync.memoize(myFunc);
+		myMemo(2, 4, 3).then(function(result) {
+			expect(result).to.equal(24);
+			done();
```
Comment (oppenlander) [inline test/tests.js]: Nest this, a couple times with different/same arguments, so we can check if it's actually being memoized.

File: test/tests.js [hunk 5]
```diff
@@ -1053,6 +1070,30 @@ describe('pasync', function() {
 		}).catch(done);
 	});
 
+	it('memoize', function(done) {
+		var myFunc = function(up, down, no) {
+			return Promise.resolve(up * down * no);
```
Comment (oppenlander) [inline test/tests.js]: Count the number of times this memoize function gets called

File: test/tests.js [hunk 6]
```diff
@@ -1053,6 +1070,30 @@ describe('pasync', function() {
...
+		myMemo(2, 4, 3).then(function(result) {
+			expect(result).to.equal(24);
+			done();
+		}).catch(done);
+	});
+
+	it('unmemoize', function(done) {
+		var myMemo;
+		var myFunc = function(up) {
+			return Promise.resolve(up * 2);
```
Comment (oppenlander) [inline test/tests.js]: Count the number of times this function is called.

File: test/tests.js [hunk 7]
```diff
@@ -1053,6 +1070,30 @@ describe('pasync', function() {
...
+		};
+		myMemo = pasync.memoize(myFunc);
+		var noMemo = pasync.unmemoize(myMemo);
+		noMemo(2).then(function(thing) {
+			expect(thing).to.equal(4);
```
Comment (oppenlander) [inline test/tests.js]: Nest this call, to ensure the memoize function is called all the time (since nothing should be memoized in unmemoize).

**Comment by crowelch** [comment]
[Take a look at me now](https://www.youtube.com/watch?v=_ZMUeBJOWhA)

**Pull request merged**